base_layer/core/src/base_node/states/horizon_state_sync/validators/headers.rs

```diff
@@ -0,0 +1,250 @@
+        tip_header: &BlockHeader,
+    ) -> Result<(), ValidationError>
+    {
+        let pow_algo = block_header.pow.pow_algo;
+        let target = if self.is_genesis(block_header) {
+            Difficulty::from(1)
+        } else {
+            let target_difficulties = self.fetch_target_difficulties(block_header, tip_header)?;
+
+            let constants = self.rules.consensus_constants();
+            get_target_difficulty(
+                target_difficulties,
+                constants.get_difficulty_block_window() as usize,
+                constants.get_diff_target_block_interval(),
+                constants.min_pow_difficulty(pow_algo),
+                constants.get_difficulty_max_block_interval(),
+            )
+            .map_err(|err| {
+                error!(
+                    target: LOG_TARGET,
+                    "Validation could not get target difficulty: {}", err
+                );
+                ValidationError::BlockHeaderError(BlockHeaderValidationError::ProofOfWorkError(
+                    PowError::InvalidProofOfWork,
+                ))
+            })?
+        };
+
+        if block_header.pow.target_difficulty != target {
+            warn!(
+                target: LOG_TARGET,
+                "Recorded header target difficulty was incorrect: (got = {}, expected = {})",
+                block_header.pow.target_difficulty,
+                target
+            );
+            return Err(ValidationError::BlockHeaderError(
+                BlockHeaderValidationError::ProofOfWorkError(PowError::InvalidTargetDifficulty),
+            ));
+        }
+
+        let achieved = block_header.achieved_difficulty();
+        if achieved < target {
+            warn!(
+                target: LOG_TARGET,
+                "Proof of work for {} was below the target difficulty. Achieved: {}, Target:{}",
+                block_header.hash().to_hex(),
+                achieved,
+                target
+            );
+            return Err(ValidationError::BlockHeaderError(
+                BlockHeaderValidationError::ProofOfWorkError(PowError::AchievedDifficultyTooLow),
+            ));
+        }
+
+        Ok(())
+    }
+
+    /// This function tests that the block timestamp is greater than the median timestamp at the specified height.
+    pub fn check_median_timestamp(
+        &self,
+        block_header: &BlockHeader,
+        tip_header: &BlockHeader,
+    ) -> Result<(), ValidationError>
+    {
+        if self.is_genesis(block_header) {
+            // Median timestamps check not required for the genesis block header
+            return Ok(());
+        }
+
+        let start_height = block_header
+            .height
+            .saturating_sub(self.rules.consensus_constants().get_median_timestamp_count() as u64);
+
+        if start_height == tip_header.height {
+            return Ok(());
+        }
+
+        let block_nums = (start_height..tip_header.height).collect();
```

Comment:
should this not be the block_header height?

Median time is checked to your previous blocks. This check will fail if you dont check the tip header.

Reply:
Good catch. Fixed